src/GPLikelihoods.jl

```diff
@@ -28,6 +28,7 @@ export Link,
     ProbitLink,
     NormalCDFLink,
     SoftMaxLink
+export DefaultQuadrature, Analytic, GaussHermite, MonteCarlo
```

Comment:
Because `Analytic` and `MonteCarlo` are technically not quadrature methods? (correct me if I am wrong)
To go with your remark `DefaultQuadrature` should be called `DefaultIntegrationMethod` I guess.

I don't see the point of going with the SciML interface, cause we don't need one. The interface approaches two different problems. We have cases where we want to use sample / analytic and another where we want to use quadrature.
However I don't see the point of having different quadrature methods. Gauss-Hermite is already optimal given an expectation over a Gaussian distribution. (again correct me if it's incorrect)

Reply:
But +1 on too generic names.
I guess adding an `Integration` postfix would solve the problem.

Reply:
> However I don't see the point of having different quadrature methods.

Well, in the sense of Quadrature.jl also eg Monte Carlo methods (and Quasi Monte Carlo methods: https://github.com/SciML/Quadrature.jl/issues/11) are "quadrature" algorithms. So I guess the main issue are not the algorithms but that Quadrature is written for problems of integrating a generic function `f`, without any decomposition into `g(x) * d(x)` where `d` is a density.

> Gauss-Hermite is already optimal given an expectation over a Gaussian distribution.

Optimal for a given tolerance? And optimal among Gauss quadrature methods, general quadrature rules, or general numerical integration algorithms?

Reply:
Yeah I agree the names are too generic, it's just what we had in ApproximateGPs before. Happy to change it before we merge it.

I don't remember having come across SciML/Quadrature, your suggestion might have simply slipped through some cracks, sorry! It would be great to re-use that, but if as theo says the interface doesn't quite work for us, then that's no good.

Iirc, nth-order Gauss-Hermite is exact for integration of nth-order polynomials under a Gaussian density. It's not necessarily optimal. It's been a while and I've never looked into it deep enough, but in GPflow context at some point I was discussing with someone about unscented transform(?) potentially being better than Gauss-Hermite.

Reply:
Hmm unscented transform can definitely be better than Monte Carlo but I am not sure about Gauss-Hermite?

Reply:
I still don't know which kind of optimality you're referring to. I guess I've been involved in SciML for too long and hence expect from a numerical approximation method to be able to specify a desired accuracy a priori - like in DE solvers or the integration methods/packages supported by Quadrature. Optimality then just means fastest method for a given error, as illustrated in the work precision diagrams. And some methods are then eg better in the low tolerance regime and some in the high tolerance regime. I've used FastGaussQuadrature myself but this concept of having a fixed number of nodes without adapting it to your problem and desired precision goes against my numerical intuition and feels unsatisfying. And makes me wonder in what sense one can claim optimality.

(I'm fully aware though that it's nice and efficient to compute a fixed set of nodes upfront in eg an iterative optimization procedure. And it might be "good enough". But I'd like to have guarantees on "good enough" and use that to choose the smallest number of nodes for which the result is still "good enough".)

Reply:
I think the discrepancy in our understanding is in the way we use quadrature. In GPs quadrature is just a small piece and we just want something as good as possible with the least possible computations and that's exactly what FGQ is doing.
Additionally, we need to solve the quadrature problem N times per step, so starting to care about tolerance might slow things down too much.
Maybe I have the (possibly wrong) view that the likelihoods used are always quite well-behaved and FGQ is always doing the trick for this problem.

Reply:
Yeah, as I wrote, it's nice that you only need one set of precomputed nodes.

> we just want something as good as possible with the least possible computations and that's exactly what FGQ is doing.

Yeah but that's exactly my point: how do you know it's good enough AND performant if you just use eg 200 nodes? Maybe 50 would have been sufficient and you could have been much more efficient. Or maybe you're result is not very accurate but you never know. The most natural approach would be to rerun computations with different number of nodes, possibly even optimizing the number of nodes - is this done in practice?

Reply:
Yep! I agree it's bad but that's what most people do. It's not the only thing after all :) (# of inducing points etc).
That's also why I prefer augmentations :p : you don't need any quadrature :)

But again, it's generally not considered an issue because the integrand used in GPs are usually well-behaved and people just use something round like 100 nodes and are happy with it.

Reply:
@devmotion are you happy to resolve this now that @st-- has agreed that we don't need to export the names?